staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go

```diff
@@ -0,0 +1,297 @@
+type FlowSchemaConditionType string
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfiguration represents the configuration of a priority level.
+type PriorityLevelConfiguration struct {
+	metav1.TypeMeta `json:",inline"`
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// `spec` is the specification of the desired behavior of a "request-priority".
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec PriorityLevelConfigurationSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	// `status` is the current status of a "request-priority".
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status PriorityLevelConfigurationStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfigurationList is a list of PriorityLevelConfiguration objects.
+type PriorityLevelConfigurationList struct {
+	metav1.TypeMeta `json:",inline"`
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// `items` is a list of request-priorities.
+	Items []PriorityLevelConfiguration `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// PriorityLevelConfigurationSpec is specification of a priority level
+type PriorityLevelConfigurationSpec struct {
+	// `exempt` defines whether the priority level is exempted or not.  There should be at most one exempt priority level.
+	// Being exempt means that requests of that priority are not subject to concurrency limits (and thus are never queued)
+	// and do not detract from the concurrency available for non-exempt requests. In a more sophisticated system, the
+	// exempt priority level would be the highest priority level. The field is default to false and only those system
+	// preset priority level can be exempt.
+	// +optional
+	Exempt bool `json:"exempt,omitempty" protobuf:"varint,1,opt,name=exempt"`
+	// `assuredConcurrencyShares` is a positive number for a non-exempt priority level, representing the weight by which
+	// the priority level shares the concurrency from the global limit. The concurrency limit of an apiserver is divided
+	// among the non-exempt priority levels in proportion to their assured concurrency shares. Basically this produces
+	// the assured concurrency value (ACV) for each priority level:
+	//
+	//             ACV(l) = ceil( SCL * ACS(l) / ( sum[priority levels k] ACS(k) ) )
```

Comment:
I am under the impression that this design would be extendable, in particular, I think eventually we want two alternative choices here in addition to this:
* a flat percentage of throughput, i.e., 5%. The rates must not add up to more than X%, say, 80%. The rest of the throughput is divided among those configured with shares. This is useful for usages that we know will grow proportionately with the control plane.
* an exact concurrency limit, e.g., 5. This is useful for constraining usages that are fixed and do not grow with the control plane.

Of course, internally, we can convert all of these to a share value or other common format, this is just about making it easy to configure.

Please let me know if this is not the case.

Reply:
Yes, we have talked about extensions that have lower bounds and/or upper bounds and we also talked about dynamically sharing concurrency capacity between priority levels.  I thought we agreed that what we care about is the product of rate and service time, which is concurrency.

Reply:
Both of the things I mentioned were about concurrency.

Eventually, it may be interesting to figure out how to provide a concurrency limit that's > 0 but < 1. I have had at least one person ask me for a rate limit for some users/namespaces/etc. I currently think that the design is extendable in this respect and therefore we can figure it out later when there's a more concrete need; if that's not the case then it'd be good to know now.

Reply:
I think a lot of extensions are possible.

Reply:
> Of course, internally, we can convert all of these to a share value or other common format, this is just about making it easy to configure.

configuring by `shares` for the priority-level is indeed not as easy as expected in the current state.. but it's compatible w/ the legacy max-flight handler. users get migrate between the two smoothly.

> Eventually, it may be interesting to figure out how to provide a concurrency limit that's > 0 but < 1. 

yes, i believe there's many useful use-cases for extra low concurrency. e.g. for some third-party components, we don't expect too many DELETE requests from them in the production, especially un-limited deletion over pods is sort of dangerous even a `1` concurrency cannot ease the pain..